### DISCUSSION TITLE
Long-running workflow run `GITHUB_TOKEN` timeout warning

### DISCUSSION THREAD

**Discussion by trym-b** [community/community#50472] [Actions]
### Select Topic Area

Product Feedback

### Body

We recently discovered that long-running workflow jobs (runs for more than 24 hours) now prints the warning on the workflow summary page:

```
<Job name>
Unable to extend GITHUB_TOKEN expiration time due to: GITHUB_TOKEN has expired.
```

![image](https://user-images.githubusercontent.com/22957880/226274799-140e735a-3113-4fe0-aa74-3dfe1f6da84d.png)

It is pretty easy to reproduce it, as you can simply run the following workflow:

```yaml
name: "Workflow with long-running job"
on: push
jobs:
    main:
        name: "Long running job"
        runs-on: [self-hosted, Linux]
        timeout-minutes: 3000
        steps:
            - name: "Long running step"
              run: "sleep 48h"
```

And the more jobs that run after 24 hours have passed, the more warnings the user will get. This is a problem, as if you have a workflow that runs a great many jobs that all run still run after 24 hours after workflow run start, they will generate a warning each.

I created a Github actions ticket to ask about what we could do to work around this issue, and I was told by the helpful Github employee that:

> The warning(s) are expected to report in the logs due to the 24 hours timeout for GITHUB_TOKEN and the only way to avoid this is to extend the lifespan of GITHUB_TOKEN but this cannot be considered as an improvement because there are possibilities of token leak with the long lived access token.

and when asked about extending the `GITHUB_TOKEN`, I got the answer

> I believe I had missed to clarify this information, the validity of GITHUB_TOKEN secret cannot be extended, we had escalations/feature requests reported earlier for the same request with our Engineering Team but they were not considered as an improvement due to the risk of secret leak.

These are some of the alternatives I have come up with, but neither may be sufficient:
1. Make it possible to suppress the warning, preferably per workflow
2. Extend the life-cycle of the `GITHUB_TOKEN`, but this might not be the smartest seen from a security perspective.
3. Remove the warning altogether, and rely on the API error messages to show that the token has expired/is not valid.
4. Display only one warning per workflow run, instead of one warning per job

The suggestions above might be impossible or difficult to implement, as I have not even taken into account how this is supposed to work when a job is re-run after the workflow has been completed.

**Comment by brianjmurrell** [comment]
Token leak?  But only if it lives longer than 24h?  How does that make any sense?

Either the token can leak or it cannot.  What does time have to do with it?  If it might leak after 24h why not 12h or 1h?

**Comment by stoyanov-x** [comment]
@trym-b or @brianjmurrell Have you had any success in working around this 24-hour token lifetime limit? Can you share your mileage?

**Reply by trym-b** [reply]
Are you trying to use the token after a job has been running for more than 24 hours?

One alternative is to add a manually created Github token that contains the permissions need as a repository secret, and then use that token instead. But you then need to consider the drawbacks of managing the token yourself.
